Add rocc test/example

NathanTP opened the pull request:
Here is the promised rocc example. We will need to add documentation to a different repo (currently marshal docs live in firesim). Some specific things I'd like feedback on:

1. **rocc.h**: I got this file from Howie who got it from someone who presumably got it from Schuyler. After committing and documenting this, it will become a de-facto standard. If anyone wants to take a stab at making a better one, be my guest. More generally, we can discuss if a template like this is really the best approach, or if we should provide more tooling to automate the process.

1. **README/Docs**: I don't claim to actually know how any of this works. I wrote the readme to the best of my ability, but someone like Howie or Albert might do a better job (or at least double check my explanations).

1. **using dummy_rocc**: I wanted to keep the example short and to the point, so I used the default 'dummy_rocc' accelerator that comes with spike. Most people will need to create a golden model and run a custom spike. The Sha3 example from the chipyard tutorial shows all that. I think after the tutorial, we might include that by default and use it as an additional example.

zhemao commented on test/rocc/util.h:
```diff
@@ -0,0 +1,135 @@
+// See LICENSE for license details.
```
Is any of the code in this header actually used? It's included in syscalls.c, but none of the code here appears to be used anywhere else.

NathanTP replied:
static_assert is used somewhere. The binary fails to link without including util.h due to a missing "static_assert".

alonamid commented on test/rocc/README:
```diff
@@ -0,0 +1,28 @@
+
+Run in Marshal:
+    ./marshal build test/rocc.json
+    ./marshal launch -s test/rocc.json
```
"Run in Marshal" here is "Run in FireSim"?
(Assuming we're moving marshal up on level in the hierarchy, this is not trivial anymore. Need to differentiate between running in FireSim, running in Verilator, or running in QEMU/Spike)

NathanTP replied:
No, this is "Run using Marshal" (using spike). I don't include instructions on running firesim in this readme, I suppose I could.

alonamid replied:
Then maybe make this "Run in Marshal using Spike"?
(I think it would be useful to have instructions for bare-metal on FireSim and Verilator. We have the Explicit Verilator instructions in the software-simulation section, but then it's not clear how it relates to Marshal)

alonamid commented on test/rocc/README:
```diff
@@ -0,0 +1,28 @@
+    ./marshal build test/rocc.json
+    ./marshal launch -s test/rocc.json
+
+or:
```
What is the difference between this option and the previous option?

NathanTP replied:
I'll clarify in the README. The first one runs the test but doesn't validate the outputs. The 'test' command automatically builds and validates outputs.

alonamid replied:
Validate against what?

NathanTP replied:
This readme assumes you've read the documentation of FireMarshal. I don't want to repeat everything here. This is just a minimal description. It should explain how building and running a RoCC workload is different than other workloads. These one-liners are here just so you can copy/paste if you forgot the exact syntax.

alonamid replied:
So it might be better to remove the last line (test) in order to not confuse anyone (if we're expecting people to have read the full documentation before hand, then they will know the `test` feature exist, and they don't need it in the readme).

NathanTP replied:
I took another stab at the README. Let me know if it's more clear now.

NathanTP commented:
I made a few updates, this is ready for re-review:
- I updated to the latest dev branch
- Removed the committed generated files (and updated the gitignore)
- Updated the README

alonamid commented on test/rocc/rocc-test.c:
```diff
@@ -0,0 +1,48 @@
+#include <stdint.h>
+#include "rocc.h"
+
+// Old custom format:
```
I'm not sure I understood this comment

NathanTP replied:
Whoops! That was a note-to-self from when I was converting this code to Schuylers macros. I've removed it.
